proposals/p3848.md

```diff
@@ -0,0 +1,587 @@
+-   [Function Fields](#function-fields)
+-   [Copy Semantics](#copy-semantics)
+-   [Self and Recursion](#self-and-recursion)
+    -   [Alternative Considered: Recursive Self](#alternative-considered-recursive-self)
+-   [Rationale](#rationale)
+
+<!-- tocstop -->
+
+## Abstract
+
+This document proposes a path forward to add lambdas to Carbon. It further
+proposes augmenting function declarations to create a more continuous syntax
+between the two categories of functions. In short, both lambdas and function
+declarations will be introduced with the `fn` keyword. The presence of a name
+distinguishes a function declaration from a lambda expression, and the rest of
+the syntax applies to both kinds. See [Syntax Overview](#syntax-overview) for
+more.
+
+## Syntax Overview
+
+**Proposal**: A largely continuous syntax between lambdas and function
+declarations.
+
+At a high level, lambdas and function declarations will look like the following.
+
+```
+let lambda1: auto = fn => T.Make();
+
+let lambda2: auto = fn []() -> T { return T.Make(); };
+
+fn FunctionDeclaration1 => T.Make();
+
+fn FunctionDeclaration2[]() -> T { return T.Make(); }
+```
+
+There are functions which return an expression, such that the return type is
+`auto`.
+
+```
+let lambda: auto = fn => T.Make();
+
+fn FunctionDeclaration => T.Make();
+```
+
+And there are functions with an explicit return type that provide a body of
+statements.
+
+```
+let lambda: auto = fn -> T { return T.Make(); };
+
+fn FunctionDeclaration -> T { return T.Make(); }
+```
+
+There are even functions that provide a body of statements but no return value.
+
+```
+let lambda: auto = fn { Print(T.Make()); };
+
+fn FunctionDeclaration { Print(T.Make()); }
+```
+
+Functions support [captures](#function-captures), [fields](#function-fields) and
+deduced parameters in the square brackets. In addition, `self: Self` or
+`addr self: Self*` can be added to the square brackets of function declarations
+that exist inside class or interface definitions.
+
+```
+fn Foo(x: X) {
+  let lambda: auto = fn [var x, var y: i32 = 0] { Print(x, y); };
+
+  fn FunctionDeclaration[var x, var y: i32 = 0] { Print(x, y); }
+}
+```
+
+Functions also support so-called
+["positional parameters"](#positional-parameters) that are defined at their
+point of use using a dollar sign and a non-negative integer. They are implicitly
+of type `auto`.
+
+```
+fn Foo() {
+  let lambda: auto = fn { Print($0); };
+
+  fn FunctionDeclaration { Print($0); }
+}
+```
+
+Of course, functions can also have named parameters, but a single function can't
+have both named and positional parameters.
+
+```
+fn Foo() {
+  let lambda: auto = fn(v: auto) { Print(v); };
+
+  fn FunctionDeclaration(v: auto) { Print(v); }
+}
+```
+
+And deduced parameters.
+
+```
+fn Foo() {
+  let lambda: auto = fn[T:! Printable](t: T) { Print(t); };
+
+  fn FunctionDeclaration[T:! Printable](t: T) { Print(t); }
+}
+```
+
+### Succinctly
```

Comment:
Not sure what this section heading means....

(I fear it may be ... too succinct? I'll see myself out.)